src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/LanguageServices/RegexPatternDetector.cs

```diff
@@ -0,0 +1,349 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Text.RegularExpressions;
+using System.Threading;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.VirtualChars;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.EmbeddedLanguages.RegularExpressions.LanguageServices
+{
+    /// <summary>
+    /// Helper class to detect regex pattern tokens in a document efficiently.
+    /// </summary>
+    internal class RegexPatternDetector
+    {
+        private const string _patternName = "pattern";
+
+        private static readonly ConditionalWeakTable<SemanticModel, RegexPatternDetector> _modelToDetector =
+            new ConditionalWeakTable<SemanticModel, RegexPatternDetector>();
+
+        private readonly RegexEmbeddedLanguage _language;
+        private readonly SemanticModel _semanticModel;
+        private readonly INamedTypeSymbol _regexType;
+        private readonly HashSet<string> _methodNamesOfInterest;
+
+        /// <summary>
+        /// Helps match patterns of the form: language=regex,option1,option2,option3
+        /// 
+        /// All matching is case insensitive, with spaces allowed between the punctuation.
+        /// 'regex' or 'regexp' are both allowed.  Option values will be or'ed together
+        /// to produce final options value.  If an unknown option is encountered, processing
+        /// will stop with whatever value has accumulated so far.
+        /// 
+        /// Option names are the values from the <see cref="RegexOptions"/> enum.
+        /// </summary>
+        private static readonly Regex s_languageCommentDetector = 
+            new Regex(@"language\s*=\s*regex(p)?((\s*,\s*)(?<option>[a-zA-Z]+))*",
+                RegexOptions.ExplicitCapture | RegexOptions.IgnoreCase | RegexOptions.Compiled);
```

Comment:
when this feature is checked in, this actual regex string will be so much more readable :)